contents/javascript/Javascript-promises.md

```diff
@@ -0,0 +1,104 @@
+# Javascript: Promises
+
+Author: Daniel Berzin Chua
+
+In Javascript, there are occasions where a method call may not resolve instantly and as such, the code may not run in the sequence that you would expect due this asynchronous nature. It can be especially confusing for beginners learning Javascript.
+
+A pertinent example would be HTTP requests. When making a HTTP request, the information from the server is not instantly returned as there is a round trip time that involves the time taken for the request to be sent, and for the response from the server to be sent back to the client. In this time, Javascript would not necessarily wait for the response to be received, but instead execute other functions first, even if those functions depend on the result of that HTTP response.
+
+In the code snippet below, `console.log(price)` would return undefined as the asynchronous `retrievePrice()` function has not finished executing.
+
+```
+var price;
+price = retrievePrice("https://www.amazon.com/gp/product/B009S331VU/");
+console.log(price);
+
+```
+
+As such, it can be difficult to predict how your Javascript code executes as asynchronous methods such as the above would not necessarily execute in the sequence that is desired. 
```

Comment:
"it can be difficult to predict how" - I think maybe it's better to not describe as "predict" (as though it's a random variable). Perhaps we can say that retrievePrice is non-blocking (this is keyword) and the I/O operation would be done in a different thread (depends on platform implementation) etc.